### PR TITLE
feat: [CI-4534]: add os field for k8s build infra

### PR DESCRIPTION
**Pull request by d1wilko**
##### Summary:

Add OS field to k8s build infra tab

Stop users from choosing build&push build steps if windows OS is chosen for k8s infra

##### Jira Links:

https://harness.atlassian.net/browse/CI-4534

##### Screenshots:

image

You can use the following comments to re-trigger PR Checks

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier - `fix prettier`

### CODE REVIEW

**Comment by d1wilko** [comment]
retrigger coverage

File: src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx [hunk 1]
```diff
@@ -865,6 +865,48 @@ export function StageInputSetFormInternal({
...
+                    'spec.spec.os',
+                    deploymentStageTemplate.infrastructure
+                  ) ? (
+                    renderMultiTypeInputWithAllowedValues({
+                      name: `${namePath}infrastructure.spec.os`,
+                      tooltipId: 'os',
+                      labelKey: osLabel,
+                      placeholderKey: osLabel,
+                      fieldPath: 'spec.os',
+                      allowedTypes: [MultiTypeInputType.FIXED]
```
Comment (mtran7) [inline src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx:879]: should this also support expression and runtime input?

File: src/modules/75-ci/components/PipelineStudio/BuildInfraSpecifications/BuildInfraSpecifications.tsx [hunk 2]
```diff
@@ -1429,7 +1461,19 @@ export default function BuildInfraSpecifications({ children }: React.PropsWithCh
           dropCapabilities: yup.lazy(value => validateUniqueList({ value, getString })),
           tolerations: yup.lazy(value =>
             validateUniqueList({ value, getString, uniqueKey: 'key', stringKey: 'pipeline.ci.validations.keyUnique' })
-          )
+          ),
+          os: yup
+            .string()
+            .test(
+              'OS required only for New configuration',
+              getString('fieldRequired', { field: 'OS' }) || '',
```
Comment (mtran7) [inline src/modules/75-ci/components/PipelineStudio/BuildInfraSpecifications/BuildInfraSpecifications.tsx:1469]: Use strings.en.yaml reference for oS

File: src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx (hunk 1 above)
Comment (d1wilko) [reply src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx:879]: no, the underlying field supports fixed and runtime, but not expression

this is basically where the runtime input get added, so we shouldn't input another runtime value

File: src/modules/70-pipeline/components/PipelineStudio/StepPalette/StepPalette.tsx [hunk 3]
```diff
@@ -11,18 +11,33 @@ import { Color } from '@harness/design-system'
...
+import type { K8sDirectInfraYaml } from 'services/ci'
+import { OsTypes } from '@pipeline/utils/constants'
 import type { AbstractStepFactory, StepData as FactoryStepData } from '../../AbstractSteps/AbstractStepFactory'
 
 import { iconMapByName } from './iconMap'
 import css from './StepPalette.module.scss'
 
+// do not show certain steps if build infra is Windows k8s
+const stepShouldDisplay = (buildInfraType: string, os: OsTypes, stepData: StepData): boolean => {
+  return buildInfraType === 'KubernetesDirect' &&
```
[the remark this answers is not in the document]
Comment (d1wilko) [reply src/modules/70-pipeline/components/PipelineStudio/StepPalette/StepPalette.tsx:34]: I will change this - but the reason I keep this as a string was that it is simply fluke that `StepType.KubernetesDirect` matches the build infra type "KubernetesDirect" - technically step types and build infra types have no relation - if that makes sense

File: src/modules/75-ci/components/PipelineStudio/BuildInfraSpecifications/BuildInfraSpecifications.tsx (hunk 2 above)
Comment (d1wilko) [reply src/modules/75-ci/components/PipelineStudio/BuildInfraSpecifications/BuildInfraSpecifications.tsx:1469]: nice catch - changing

**Review by kapilkumawat86 — CHANGES_REQUESTED** [review]
don't write this logic in UI 
image
move to BE

File: src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx (hunk 1 above)
Comment (mtran7) [reply src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx:879]: oh right! Got it

**Pull request closed without merging**